File: ios/CareLog/CareLog/Dashboard/ViewModels/DashboardViewModel.swift
import Foundation
import SwiftUI
import Combine

/// ViewModel for patient dashboard.
@MainActor
final class DashboardViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var patientName: String?
    @Published var lastValues: [VitalType: String] = [:]
    @Published var pendingSyncCount: Int = 0
    @Published var missedReminder: VitalType?
    @Published var isLoading: Bool = false

    // MARK: - Dependencies

    private let authService: AuthService
    private let localFHIRRepository: LocalFHIRRepository
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init(authService: AuthService = .shared, localFHIRRepository: LocalFHIRRepository = .shared) {
        self.authService = authService
        self.localFHIRRepository = localFHIRRepository

        Task {
            await loadDashboardData()
        }
        observePendingSyncCount()
    }

    // MARK: - Data Loading

    func loadDashboardData() async {
        isLoading = true
        defer { isLoading = false }

        // Load patient info
        if let user = await authService.getCurrentUser() {
            patientName = user.name

            // Load last values for each vital type
            if let patientId = user.linkedPatientId {
                await loadLastValues(patientId: patientId)
            }
        }
    }

    private func loadLastValues(patientId: String) async {
        var values: [VitalType: String] = [:]

        // Blood Pressure
        if let obs = await localFHIRRepository.getLatestObservation(
            patientId: patientId,
            type: .bloodPressure
        ) {
            if let systolic = obs.components?.first(where: { $0.type == .systolicBP })?.value,
               let diastolic = obs.components?.first(where: { $0.type == .diastolicBP })?.value {
                values[.bloodPressure] = "\(Int(systolic))/\(Int(diastolic)) mmHg"
            }
        }

        // Glucose
        if let obs = await localFHIRRepository.getLatestObservation(
            patientId: patientId,
            type: .bloodGlucose
        ), let value = obs.value {
            values[.glucose] = "\(Int(value)) mg/dL"
        }

        // Temperature
        if let obs = await localFHIRRepository.getLatestObservation(
            patientId: patientId,
            type: .bodyTemperature
        ), let value = obs.value {
            values[.temperature] = String(format: "%.1f\u{00B0}F", value)
        }

        // Weight
        if let obs = await localFHIRRepository.getLatestObservation(
            patientId: patientId,
            type: .bodyWeight
        ), let value = obs.value {
            values[.weight] = String(format: "%.1f kg", value)
        }

        // Pulse
        if let obs = await localFHIRRepository.getLatestObservation(
            patientId: patientId,
            type: .heartRate
        ), let value = obs.value {
            values[.pulse] = "\(Int(value)) bpm"
        }

        // SpO2
        if let obs = await localFHIRRepository.getLatestObservation(
            patientId: patientId,
            type: .oxygenSaturation
        ), let value = obs.value {
            values[.spO2] = "\(Int(value))%"
        }

        lastValues = values
    }

    private func observePendingSyncCount() {
        Task {
            for await count in localFHIRRepository.pendingObservationCountStream() {
                self.pendingSyncCount = count
            }
        }
    }

    // MARK: - Actions

    func dismissReminder() {
        missedReminder = nil
    }

    func setMissedReminder(_ vitalType: VitalType) {
        missedReminder = vitalType
    }

    func refreshData() {
        Task {
            await loadDashboardData()
        }
    }
}

// MARK: - Auth Service Protocol (for dependency injection)

protocol AuthServiceProtocol {
    func getCurrentUser() async -> UserInfo?
}

struct UserInfo {
    let name: String
    let linkedPatientId: String?
}

/// Placeholder AuthService - replace with actual implementation
class AuthService: AuthServiceProtocol {
    static let shared = AuthService()

    func getCurrentUser() async -> UserInfo? {
        // TODO: Implement actual auth service
        return nil
    }
}

File: android/app/src/main/java/com/carelog/ui/dashboard/DashboardViewModel.kt
package com.carelog.ui.dashboard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.carelog.auth.AuthRepository
import com.carelog.fhir.client.ObservationType
import com.carelog.fhir.repository.LocalFhirRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for patient dashboard.
 */
@HiltViewModel
class DashboardViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val localFhirRepository: LocalFhirRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(DashboardUiState())
    val uiState: StateFlow<DashboardUiState> = _uiState.asStateFlow()

    init {
        loadDashboardData()
        observePendingSyncCount()
    }

    private fun loadDashboardData() {
        viewModelScope.launch {
            // Load patient info
            val user = authRepository.getCurrentUser()
            _uiState.update { it.copy(patientName = user?.name) }

            // Load last values for each vital type
            user?.linkedPatientId?.let { patientId ->
                loadLastValues(patientId)
            }
        }
    }

    private suspend fun loadLastValues(patientId: String) {
        val lastValues = mutableMapOf<VitalType, String>()

        // Blood Pressure
        localFhirRepository.getLatestObservation(patientId, ObservationType.BLOOD_PRESSURE)?.let { obs ->
            obs.components?.let { components ->
                val systolic = components.find { it.type == ObservationType.SYSTOLIC_BP }?.value?.toInt()
                val diastolic = components.find { it.type == ObservationType.DIASTOLIC_BP }?.value?.toInt()
                if (systolic != null && diastolic != null) {
                    lastValues[VitalType.BLOOD_PRESSURE] = "$systolic/$diastolic mmHg"
                }
            }
        }

        // Glucose
        localFhirRepository.getLatestObservation(patientId, ObservationType.BLOOD_GLUCOSE)?.let { obs ->
            obs.value?.let { value ->
                lastValues[VitalType.GLUCOSE] = "${value.toInt()} mg/dL"
            }
        }

        // Temperature
        localFhirRepository.getLatestObservation(patientId, ObservationType.BODY_TEMPERATURE)?.let { obs ->
            obs.value?.let { value ->
                lastValues[VitalType.TEMPERATURE] = String.format("%.1f°F", value)
            }
        }

        // Weight
        localFhirRepository.getLatestObservation(patientId, ObservationType.BODY_WEIGHT)?.let { obs ->
            obs.value?.let { value ->
                lastValues[VitalType.WEIGHT] = String.format("%.1f kg", value)
            }
        }

        // Pulse
        localFhirRepository.getLatestObservation(patientId, ObservationType.HEART_RATE)?.let { obs ->
            obs.value?.let { value ->
                lastValues[VitalType.PULSE] = "${value.toInt()} bpm"
            }
        }

        // SpO2
        localFhirRepository.getLatestObservation(patientId, ObservationType.OXYGEN_SATURATION)?.let { obs ->
            obs.value?.let { value ->
                lastValues[VitalType.SPO2] = "${value.toInt()}%"
            }
        }

        _uiState.update { it.copy(lastValues = lastValues) }
    }

    private fun observePendingSyncCount() {
        viewModelScope.launch {
            localFhirRepository.getPendingObservationCount()
                .combine(localFhirRepository.getPendingDocumentCount()) { obsCount, docCount ->
                    obsCount + docCount
                }
                .collect { count ->
                    _uiState.update { it.copy(pendingSyncCount = count) }
                }
        }
    }

    fun dismissReminder() {
        _uiState.update { it.copy(missedReminder = null) }
    }

    fun setMissedReminder(vitalType: VitalType) {
        _uiState.update { it.copy(missedReminder = vitalType) }
    }

    fun refreshData() {
        loadDashboardData()
    }
}

/**
 * UI state for dashboard.
 */
data class DashboardUiState(
    val patientName: String? = null,
    val lastValues: Map<VitalType, String> = emptyMap(),
    val pendingSyncCount: Int = 0,
    val missedReminder: VitalType? = null,
    val isLoading: Boolean = false
)
